src/sagemaker_containers/_mpi.py

```diff
@@ -0,0 +1,280 @@
+import argparse
+import os
+import shlex
+import socket
+import stat
+import subprocess
+import sys
+import textwrap
+import time
+
+from retrying import retry
+
+import sagemaker_containers
+from sagemaker_containers import _logging
+from sagemaker_containers._mapping import to_cmd_args
+from sagemaker_containers._timeout import timeout
+
+logger = _logging.get_logger()
+
+_MPI_SCRIPT_TEMPLATE = """#!/usr/bin/env bash
+touch %s
+%s
+EXIT_CODE=$?
+touch %s
+exit ${EXIT_CODE}
+"""
+
+# MPI files.
+_MPI_SCRIPT = "/mpi_script.sh"
+_MPI_IS_RUNNING = "/mpi_is_running"
+_MPI_IS_FINISHED = "/mpi_is_finished"
+_CHANGE_HOSTNAME_LIBRARY = "/libchangehostname.so"
+
+# MPI Configurations
+SAGEMAKER_MPI_NUM_PROCESSES_PER_HOST = "sagemaker_mpi_num_of_processes_per_host"
+SAGEMAKER_MPI_CUSTOM_MPI_OPTIONS = "sagemaker_mpi_custom_mpi_options"
+
+
+def _change_hostname(current_host):  # type: (str) -> None
+    """Compiles a shared library to correct the behavior of the gethostname system call,
+        which OpenMPI depends on.
+    Args:
+        current_host (str): name of the current host, such as algo-1, algo-2, etc.
+    """
+    os.system("/change-hostname.sh {}".format(current_host))
```

Comment:
Yep its a pain point.. other option we can do is to pacakge them along with the `sagemaker-containers` so any container that needs to use MPI will have them as they would anyway need `sagemaker-containers`

Reply:
You could also write `change_hostname.c` in the same way `mpi_script.sh` is written, and compile it with `subprocess`. This would simplify things a bit -- `change-hostname.sh` wouldn't be needed at all if you write `change_hostname.c` with the real hostname instead of placeholder.